### PR TITLE
NEWTS-115: Cassandra driver upgrade

### PR DESCRIPTION
**Pull request by j-white**
JIRA: https://opennms.atlassian.net/browse/NEWTS-115

Upgrades the Cassandra driver to the latest (4.15.0) and deals with all of the fall out.

### CODE REVIEW

File: cassandra/storage/src/main/java/org/opennms/newts/persistence/cassandra/ConcurrentResultWrapper.java [hunk 1]
```diff
@@ -31,25 +35,54 @@ public class ConcurrentResultWrapper implements Iterator<Row> {
...
             }
         });
     }
 
+    /**
+     * Convert the result-set to a blocking iterator like the previous API did
+     *
+     * The iterator blocks when the next page is loading, but at least it only loads on page at a time
+     *
+     * This is a cry for help
```
Comment (j-white) [inline cassandra/storage/src/main/java/org/opennms/newts/persistence/cassandra/ConcurrentResultWrapper.java:60]: Yes, it is

File: cassandra/common/src/main/java/org/opennms/newts/cassandra/CassandraSessionImpl.java [hunk 2]
```diff
@@ -15,115 +15,144 @@
...
 
-        final PoolingOptions poolingOptions = new PoolingOptions();
         if (coreConnectionsPerHost != null) {
-            LOG.debug("Using {} core connections per host.", coreConnectionsPerHost);
-            poolingOptions.setCoreConnectionsPerHost(HostDistance.LOCAL, coreConnectionsPerHost)
-                    .setCoreConnectionsPerHost(HostDistance.REMOTE, coreConnectionsPerHost);
+            LOG.warn("Ignoring core connections per host value of: {}. " +
+                    "Use 'max-connections-per-host' only instead.", coreConnectionsPerHost);
+
         }
```
Comment (RangerRick) [inline cassandra/common/src/main/java/org/opennms/newts/cassandra/CassandraSessionImpl.java:101]: Then why keep it at all? It's 3.0.0, it's reasonable to expect users to switch to a new signature.

File: cassandra/common/src/main/java/org/opennms/newts/cassandra/CassandraSessionImpl.java [hunk 3]
```diff
@@ -15,115 +15,144 @@
...
-        Builder builder = Cluster
-                .builder()
-                .withPort(port)
-                .addContactPoints(hostname.split(","))
-                .withReconnectionPolicy(new ExponentialReconnectionPolicy(1000, 2 * 60 * 1000))
-                .withPoolingOptions(poolingOptions)
-                .withCompression(Compression.valueOf(compression.toUpperCase()));
+        configBuilder.withClass(DefaultDriverOption.RECONNECTION_POLICY_CLASS, ExponentialReconnectionPolicy.class);
+        configBuilder.withDuration(DefaultDriverOption.RECONNECTION_BASE_DELAY, Duration.of(1, ChronoUnit.SECONDS));
+        configBuilder.withDuration(DefaultDriverOption.RECONNECTION_MAX_DELAY, Duration.of(2, ChronoUnit.MINUTES));
```
Comment (RangerRick) [inline cassandra/common/src/main/java/org/opennms/newts/cassandra/CassandraSessionImpl.java:113]: Is this something we could conceivably want to make configurable?

File: cassandra/common/src/main/java/org/opennms/newts/cassandra/CassandraSessionImpl.java [hunk 4]
```diff
@@ -15,115 +15,144 @@
...
             LOG.info("Using username: {} and password: XXXXXXXX", username);
-            builder.withCredentials(username, password);
+            configBuilder.withString(DefaultDriverOption.AUTH_PROVIDER_USER_NAME, username);
+            configBuilder.withString(DefaultDriverOption.AUTH_PROVIDER_PASSWORD, password);
         }
 
         if (ssl) {
-            LOG.info("Enabling SSL.");
-            builder.withSSL();
+            throw new IllegalArgumentException("Driver must be configured with \"cassandra.driver-settings-file\" to enable SSL support.");
```
Comment (RangerRick) [inline cassandra/common/src/main/java/org/opennms/newts/cassandra/CassandraSessionImpl.java:123]: Here too, if an option has to be done a different way, why not get rid of it altogether?

File: cassandra/common/src/main/java/org/opennms/newts/cassandra/SchemaManager.java [hunk 5]
```diff
@@ -47,29 +51,44 @@ public class SchemaManager implements AutoCloseable {
...
             LOG.info("Using username: {} and password: XXXXXXXX", username);
-            builder.withCredentials(username, password);
+            configBuilder.withString(DefaultDriverOption.AUTH_PROVIDER_USER_NAME, username);
+            configBuilder.withString(DefaultDriverOption.AUTH_PROVIDER_PASSWORD, password);
         }
 
         if (ssl) {
-            LOG.info("Using SSL.");
-            builder.withSSL();
+            throw new IllegalArgumentException("Driver must be configured with \"cassandra.driver-settings-file\" to enable SSL support.");
```
Comment (RangerRick) [inline cassandra/common/src/main/java/org/opennms/newts/cassandra/SchemaManager.java:85]: same (I'll stop making note of specific things, but you get the idea \<g>)

File: cassandra/search/src/main/java/org/opennms/newts/cassandra/search/CassandraSearcher.java [hunk 6]
```diff
@@ -220,39 +220,58 @@ private Set<String> searchForIds(Context context, BooleanQuery query, Consistenc
...
+            for (Row row : asyncResultSet.currentPage()) {
+                metricNames.add(row.getString(Schema.C_METRICS_NAME));
+            }
+            if (asyncResultSet.hasMorePages()) {
+                // blocking
+                asyncResultSet = asyncResultSet.fetchNextPage().toCompletableFuture().get();
+            } else {
+                break;
+            }
         }
```
Comment (RangerRick) [inline cassandra/search/src/main/java/org/opennms/newts/cassandra/search/CassandraSearcher.java:275]: a `while(true)` with a `break` gives me the heebie-jeebies, but I get it

File: karaf/pom.xml [hunk 7]
```diff
@@ -42,9 +46,70 @@
...
+                <feature>aries-blueprint</feature>
+                <feature>framework</feature>
+                <feature>shell</feature>
+              </framework>
+              <features>
+                <feature>netty</feature>
+                <feature>cassandra-driver</feature>
+                <feature>fastxml-jackson</feature>
+                <feature>newts*</feature>
+              </features>
```
Comment (RangerRick) [inline karaf/pom.xml:97]: 🎉

**Review by RangerRick — APPROVED** [review]
Honestly, less terrifying than I expected.

Only real feedback I had was related to keeping old options that don't work or indicate errors. Feels like it's better to either drop them or adapt them.

File: cassandra/common/src/main/java/org/opennms/newts/cassandra/CassandraSessionImpl.java (hunk 2 above)
Comment (j-white) [reply cassandra/common/src/main/java/org/opennms/newts/cassandra/CassandraSessionImpl.java:101]: Good point - I'll revisit once I work through the configuration on the OpenNMS side

File: cassandra/common/src/main/java/org/opennms/newts/cassandra/CassandraSessionImpl.java (hunk 3 above)
Comment (j-white) [reply cassandra/common/src/main/java/org/opennms/newts/cassandra/CassandraSessionImpl.java:113]: Possibly yes. This and all the other settings can be tuned when the "driver settings file" is used - will add some docs to that effect in OpenNMS.

**Pull request merged**